src/org/traccar/protocol/GranitProtocolDecoder.java

```diff
@@ -0,0 +1,212 @@
+
+    public static void appendChecksum(ChannelBuffer buffer, int length) {
+        buffer.writeByte('*');
+        int checksum = Checksum.xor(buffer.toByteBuffer(0, length)) & 0xFF;
+        String checksumString = String.format("%02X", checksum);
+        buffer.writeBytes(checksumString.getBytes(StandardCharsets.US_ASCII));
+        buffer.writeByte('\r'); buffer.writeByte('\n');
+    }
+
+    private static void sendResponseCurrent(Channel channel, int deviceId, long time) {
+        ChannelBuffer response = ChannelBuffers.dynamicBuffer(ByteOrder.LITTLE_ENDIAN, 0);
+        response.writeBytes("BB+UGRC~".getBytes(StandardCharsets.US_ASCII));
+        response.writeShort(6); //binary length
+        response.writeInt((int) time);
+        response.writeShort(deviceId);
+        appendChecksum(response, 16);
+        channel.write(response);
+    }
+
+    private static void sendResponseArchive(Channel channel, int deviceId, int packNum) {
+        ChannelBuffer response = ChannelBuffers.dynamicBuffer(ByteOrder.LITTLE_ENDIAN, 0);
+        response.writeBytes("BB+ARCF~".getBytes(StandardCharsets.US_ASCII));
+        response.writeShort(4); //binary length
+        response.writeShort(packNum);
+        response.writeShort(deviceId);
+        appendChecksum(response, 14);
+        channel.write(response);
+    }
+
+    private static void decodeStructure(ChannelBuffer buf, Position position) {
+        short flags = buf.readUnsignedByte();
+        position.setValid(BitUtil.check(flags, 7));
+        position.set(Position.KEY_ALARM, BitUtil.check(flags, 1));
+
+        short satDel = buf.readUnsignedByte();
+        position.set(Position.KEY_SATELLITES, BitUtil.from(satDel, 4));
+
+        int pdop = BitUtil.to(satDel, 4);
+        position.set("pdop", pdop);
+
+        int lonDegrees = buf.readUnsignedByte();
+        int latDegrees = buf.readUnsignedByte();
+        int lonMinutes = buf.readUnsignedShort();
+        int latMinutes = buf.readUnsignedShort();
+        double latitude = latDegrees + (double) latMinutes / 60000;
+        double longitude = lonDegrees + (double) lonMinutes / 60000;
+        if (!BitUtil.check(flags, 4)) {
+            latitude = -latDegrees;
+        }
+        if (!BitUtil.check(flags, 5)) {
+            longitude = -longitude;
+        }
+        position.setLongitude(longitude);
+        position.setLatitude(latitude);
+
+        position.setSpeed(buf.readUnsignedByte());
+
+        int course = buf.readUnsignedByte();
+        if (BitUtil.check(flags, 6)) {
+            course = course | 0x100;
+        }
+        position.setCourse(course);
+
+        position.set(Position.KEY_DISTANCE, buf.readShort());
+
+        int analogIn1 = buf.readUnsignedByte();
+        int analogIn2 = buf.readUnsignedByte();
+        int analogIn3 = buf.readUnsignedByte();
+        int analogIn4 = buf.readUnsignedByte();
+
+        int analogInHi = buf.readUnsignedByte();
+
+        analogIn1 = analogInHi << 8 & 0x300 | analogIn1;
```

Comment:
Very odd structure.
As I understand analog DAC is 10 bit, and analogInX is 8 low bits, and two other bits in analogInHi in according place.

analogInHi[0-1] is high bits of analogIn1
analogInHi[2-3] is high bits of analogIn2
analogInHi[4-5] is high bits of analogIn3
analogInHi[6-7] is high bits of analogIn3

I just move them on right place (to 8-9) clear with `& 0x300` and `or` with analogInX